File: qd_hrms/public/js/attendance_request.js
frappe.ui.form.on("Attendance Request", {
	onload(frm) {
		if (frm.is_new() && !frm.doc.employee && frappe.session.user !== "Administrator") {
			frappe.db.get_value("Employee", { user_id: frappe.session.user }, "name").then((r) => {
				if (r.message && r.message.name) {
					frm.set_value("employee", r.message.name);
				}
			});
		}
	},

	refresh(frm) {
		frm.set_intro(
			__(
				"Attendance Correction: employee requests a change; Leave Approver or HR Approves it. Approved requests submit and update Attendance. Locked payroll periods can only be changed by HR Manager."
			),
			"blue"
		);
	},

	employee(frm) {
		frm.trigger("from_date");
	},

	from_date(frm) {
		if (!frm.doc.employee || !frm.doc.from_date) return;
		frappe.db
			.get_value("Attendance", { employee: frm.doc.employee, attendance_date: frm.doc.from_date, docstatus: 1 }, "status")
			.then((r) => {
				if (r.message && r.message.status) {
					frm.set_value("custom_qd_original_status", r.message.status);
				}
			});
	},
});

File: qd_hrms/overrides/attendance_request.py
"""Attendance Request override used as Quick Delivery Attendance Correction."""

import frappe
from frappe import _
from frappe.utils import formatdate
from hrms.hr.doctype.attendance_request.attendance_request import AttendanceRequest

PRIVILEGED_APPROVER_ROLES = {"System Manager", "HR Manager", "HR User"}


class QDAttendanceRequest(AttendanceRequest):
	def validate(self):
		super().validate()
		if not self.custom_qd_requested_status:
			frappe.throw(_("Requested Attendance Status is required."))
		self._set_original_attendance_summary()

	def before_submit(self):
		self._validate_approver()

	def get_attendance_status(self, attendance_date):
		if self.get("custom_qd_requested_status"):
			return self.custom_qd_requested_status
		return super().get_attendance_status(attendance_date)

	def _validate_approver(self):
		roles = set(frappe.get_roles())
		if roles & PRIVILEGED_APPROVER_ROLES or frappe.session.user == "Administrator":
			return
		approver = frappe.db.get_value("Employee", self.employee, "leave_approver")
		if not approver or approver != frappe.session.user:
			frappe.throw(
				_("Only the employee's configured Leave Approver or HR may approve this correction."),
				frappe.PermissionError,
			)

	def _set_original_attendance_summary(self):
		rows = frappe.get_all(
			"Attendance",
			filters={
				"employee": self.employee,
				"attendance_date": ["between", [self.from_date, self.to_date]],
				"docstatus": 1,
			},
			fields=["attendance_date", "status"],
			order_by="attendance_date",
		)
		if not rows:
			self.custom_qd_original_status = _("No submitted Attendance")
			return
		self.custom_qd_original_status = "; ".join(
			f"{formatdate(row.attendance_date)}: {row.status}" for row in rows
		)
